streams/src/main/java/org/apache/kafka/streams/processor/internals/StoreChangelogReader.java

```diff
@@ -185,55 +186,55 @@ int bufferedLimitIndex() {
     }
 
     private final static long DEFAULT_OFFSET_UPDATE_MS = Duration.ofMinutes(5L).toMillis();
-
-    private ChangelogReaderState state;
+    private static final long RESTORE_LOG_INTERVAL_MS = Duration.ofSeconds(10L).toMillis();
 
     private final Time time;
     private final Logger log;
+    private final String groupId;
     private final Duration pollTime;
     private final long updateOffsetIntervalMs;
+    private long lastUpdateOffsetTime = 0L;
+    private long lastRestoreLogTime = 0L;
 
     // 1) we keep adding partitions to restore consumer whenever new tasks are registered with the state manager;
     // 2) we do not unassign partitions when we switch between standbys and actives, we just pause / resume them;
     // 3) we only remove an assigned partition when the corresponding task is being removed from the thread.
     private final Consumer<byte[], byte[]> restoreConsumer;
     private final StateRestoreListener stateRestoreListener;
 
+    // the changelog reader needs the admin client to list end offsets and the group's committed offsets
+    private final Admin adminClient;
+
     // source of the truth of the current registered changelogs;
     // NOTE a changelog would only be removed when its corresponding task
     // is being removed from the thread; otherwise it would stay in this map even after completed
+    //
+    // this map may be concurrently accessed by different threads and hence need to be guarded
```

Comment:
I'll stop pointing them all out and let you give this class a pass for any out of date comments 🙂